keps/sig-network/3453-minimize-iptables-restore/README.md

```diff
@@ -0,0 +1,473 @@
+to 0, or when a particular Service property changes).
+
+Since, as mentioned above, there are no _expected_ cases where a
+partial restore will fail, we will have a metric for when it does, so
+we can notice that this failure mode is happening.
+
+#### Subtle Synchronization Delays
+
+The new code needs to analyze Service/EndpointSlice changes to
+determine when the service-specific chains for each service need to be
+included in the `iptables-restore` input. If this analysis is
+incorrect, it might fail to update the iptables rules for a service in
+response to certain kinds of changes (eg, when a service goes from 1
+endpoint to 0, or when a particular Service property changes), causing
+the iptables rules to be stale until they eventually get fixed in a
+future full resync.
+
+(There is no particular reason to assume that this sort of bug would
+happen; it's just something that theoretically _could_ happen.)
+
+In order to detect this, we would have to sanity-check every partial
+restore, to ensure that, if applied to the current kernel state, it
+would generate the same end result as the full restore we would
+otherwise have applied. Although we could do this with the
+`FakeIPTables` code in `pkg/util/iptables/testing`, that code is only
+intended for the sorts of small-ish test cases in the unit tests, and
+not at all optimized for handling the sort of very large rule sets
+that would exist in larger clusters (ie, the rule sets that would
+benefit the most from the partial restore feature).
+
+Additionally, the code to do this checking would end up being much
+more complicated than the code that it was actually checking, and so
+any reported failures would be as likely to be the result of a bug in
+the checking code as they would be to indicate a bug in the actual
+syncing code.
+
+## Design Details
+
+### Test Plan
+
+[X] I/we understand the owners of the involved components may require updates to
+existing tests to make this code solid enough prior to committing the changes necessary
+to implement this enhancement.
+
+##### Prerequisite testing updates
+
+This is the latest in a series of iptables kube-proxy improvements,
+and many prerequisite tests have already been added.
+
+##### Unit tests
+
+We will add unit tests of the partial-resyncing feature, to ensure
+that expected rules are updated when services and endpoints
+change. (This is already done in the current PR.)
+
+Existing coverage:
+- `k8s.io/kubernetes/pkg/proxy/iptables`: `2022-08-02` - `67%`
+
+##### Integration tests
+
+Kube-proxy is mostly tested via unit tests and e2e, not integration,
+and this is not expected to change.
+
+##### e2e tests
+
+All existing kube-proxy/Service e2e tests would be testing the new
+code. In fact, a large percentage of the non-networking e2e tests
+would also end up testing the new code. There is really no good way to
+_specifically_ test this feature in any way that wasn't redundant with
+the existing e2e tests and the feature's unit tests.
```

Comment:
ping - can you please add here that the first step for checking that would be to add it to scalability tests? It should be fairly easy to do.

Reply:
I can volunteer to help here @danwinship

Reply:
[FWIW - it should boil down to adding a similar measurement to this one:
https://github.com/kubernetes/perf-tests/blob/master/clusterloader2/pkg/measurement/common/slos/network_programming.go
]

Reply:
Added scalability test modification to Alpha criteria and moved e2e test modification to Beta criteria, in particular since [there's been debate about the right way to do it in the e2e tests](https://kubernetes.slack.com/archives/C09QZ4DQB/p1663940970704269).